### PR TITLE
Ensure upsert deletion consistency when enabled with compaction flow

### PR DESCRIPTION
**Pull request by tibrewalpratik17**
### Problem statement

In #12037, we added support for removing deleted keys and the validDocId for the deleted records after a TTL window. This ensured that deleted records would eventually be picked up by the compaction flow.
In #13342, we directly enabled compaction to take care of deleted rows. [Reverted]

In both scenarios, we could end up in a situation where an older non-deleted record for a particular key is not compacted, but the deleted record is. During a server restart, when all segments are loaded, Pinot would incorrectly mark the record as non-deleted and start returning it as a valid primary key, leading to an inconsistent state in the table.

**Example**:

Consider a primary key PK1 with records in segments S0 and S1. In S1, the record is marked as deleted. If S1 gets compacted but S0 doesn't due to threshold reasons in the upsert-compaction flow, during a server restart, the upsert-metadata-manager map would incorrectly point PK1 to S0, even though it should be considered deleted for the end-user.

**Why not use enablePreload?**
While enablePreload might prevent loading previous invalid records, it is not foolproof. Valid-doc-id snapshots are not always available due to scenarios like fetching segments from deepstore or upsert-compaction refreshing the segment. If a validDocID snapshot is missing and enablePreload is set to true, the segment loads normally, iterating through records and would potentially end-up reviving a deleted primary key with an old value with the present flow.

### Proposal

This PR aims to maintain the state of PK -> distinct-segment-count. This means tracking the number of segments in which a record exists for a given primary key. If the count is <= 1, we can compact the deleted record, ensuring that all other records in other segments are removed.

Moved the proposal to this [doc](https://docs.google.com/document/d/1e035AUawzCn1SVs33NWhlBQIHQMdFrHPErgFhIwmacY/edit) updated based on offline discussions as well.

cc @klsince @Jackie-Jiang raising this draft PR for early feedback on the proposal.

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+  @Override
+  protected void removeSegment(IndexSegment segment, MutableRoaringBitmap validDocIds) {
+    try (UpsertUtils.RecordInfoReader recordInfoReader = new UpsertUtils.RecordInfoReader(segment, _primaryKeyColumns,
+        _comparisonColumns, _deleteRecordColumn)) {
+      Iterator<RecordInfo> iterator =
+          UpsertUtils.getRecordInfoIterator(recordInfoReader, segment.getSegmentMetadata().getTotalDocs());
+      while (iterator.hasNext()) {
+        RecordInfo recordInfo = iterator.next();
+        _primaryKeyToRecordLocationMap.computeIfPresent(
+            HashUtils.hashPrimaryKey(recordInfo.getPrimaryKey(), _hashFunction), (pk, recordLocation) -> {
```
Comment (tibrewalpratik17) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:188]: self-review: process one pk strictly-once during segment removal here.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BaseTableUpsertMetadataManager.java [hunk 2]
```diff
@@ -81,6 +81,8 @@ public void init(TableConfig tableConfig, Schema schema, TableDataManager tableD
         .setPartialUpsertHandler(partialUpsertHandler).setEnableSnapshot(enableSnapshot).setEnablePreload(enablePreload)
         .setMetadataTTL(metadataTTL).setDeletedKeysTTL(deletedKeysTTL).setConsistencyMode(_consistencyMode)
         .setUpsertViewRefreshIntervalMs(upsertViewRefreshIntervalMs).setTableIndexDir(tableIndexDir)
+        .setDropOutOfOrderRecord(upsertConfig.isDropOutOfOrderRecord())
+        .setEnableConsistentDeletes(upsertConfig.isEnableConsistentDeletes())
```
Comment (ankitsultana) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BaseTableUpsertMetadataManager.java:85]: Why is there a separate config for this? Shouldn't we use the new metadata manager whenever deletions are enabled?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+
+/**
+ * Implementation of {@link PartitionUpsertMetadataManager} that is backed by a {@link ConcurrentHashMap} and ensures
+ * consistent deletions. This should be used when the table is configured with 'consistentDeletes' set to true.
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+@ThreadSafe
+public class ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes
```
Comment (ankitsultana) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:52]: What's the main reason for a new metadata manager? Is it because we are adding a new int to RecordLocation, and there are perf concerns?

Comment (tibrewalpratik17) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:52]: > What's the main reason for a new metadata manager? Is it because we are adding a new int to RecordLocation, and there are perf concerns?

Yes the main reason is using a new int to RecordLocation which we discussed will add 16bytes per key(unsigned) and for supporting 100M per host can easily add 1600MB. We thought of making RecordLocation an interface as well to achieve this but then the metadataManager class will become a bit incomprehensible with lots of if-else's.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BaseTableUpsertMetadataManager.java (hunk 2 above)
Comment (tibrewalpratik17) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BaseTableUpsertMetadataManager.java:85]: So folks can enable `deletedKeysTTL` without enabling compaction as well. This config is only required when you are enabling deletedKeysTTL + compaction as this entire PR ensures data consistency for this specific scenario.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+    extends BasePartitionUpsertMetadataManager {
+
+  @VisibleForTesting
+  final ConcurrentHashMap<Object, ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.RecordLocation>
+      _primaryKeyToRecordLocationMap = new ConcurrentHashMap<>();
+  // Used to initialize a reference to previous row for merging in partial upsert
+  private final LazyRow _reusePreviousRow = new LazyRow();
+  private final Map<String, Object> _reuseMergeResultHolder = new HashMap<>();
+
+  public ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes(String tableNameWithType, int partitionId,
```
Comment (ankitsultana) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:62]: @Jackie-Jiang : just wanted to confirm, you folks are okay with creating a new metadata manager for handling deletes?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BaseTableUpsertMetadataManager.java (hunk 2 above)
Comment (tibrewalpratik17) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BaseTableUpsertMetadataManager.java:85]: Yes but won't it become non-intuitive for users to put entire `metadataManagerClass` instead of just enabling a boolean flag. Plus IMO we might make a call in future to consolidate this class with the original class and so that will make all existing tables using this `metadataManagerClass` as backward-incompatible. Better to not strictly depend on className to enable this feature.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+
+    public int getDocId() {
+      return _docId;
+    }
+
+    public Comparable getComparisonValue() {
+      return _comparisonValue;
+    }
+
+    public int getDistinctSegmentCount() {
```
[the remark this answers is not in the document]
Comment (tibrewalpratik17) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:369]: we are checking this condition in `doRemoveExpiredPrimaryKeys` that `getDistinctSegmentCount <= 1`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/TableUpsertMetadataManagerFactory.java [hunk 6]
```diff
@@ -29,16 +29,15 @@
...
-  }
-
-  private static final Logger LOGGER = LoggerFactory.getLogger(TableUpsertMetadataManagerFactory.class);
   public static final String UPSERT_DEFAULT_METADATA_MANAGER_CLASS = "default.metadata.manager.class";
   public static final String UPSERT_DEFAULT_ENABLE_SNAPSHOT = "default.enable.snapshot";
   public static final String UPSERT_DEFAULT_ENABLE_PRELOAD = "default.enable.preload";
-
   public static final String UPSERT_DEFAULT_ALLOW_PARTIAL_UPSERT_CONSUMPTION_DURING_COMMIT =
       "default.allow.partial.upsert.consumption.during.commit";
+  private static final Logger LOGGER = LoggerFactory.getLogger(TableUpsertMetadataManagerFactory.class);
```
[the remark this answers is not in the document]
Comment (tibrewalpratik17) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/TableUpsertMetadataManagerFactory.java:37]: answered here - https://github.com/apache/pinot/pull/13347/files#r1701487051

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+    if (numKeys > 0) {
+      _logger.warn("Found {} primary keys in the wrong segment when adding segment: {}", numKeys, segmentName);
+      _serverMetrics.addMeteredTableValue(_tableNameWithType, ServerMeter.UPSERT_KEYS_IN_WRONG_SEGMENT, numKeys);
+    }
+  }
+
+  @Override
+  protected void addSegmentWithoutUpsert(ImmutableSegmentImpl segment, ThreadSafeMutableRoaringBitmap validDocIds,
+      @Nullable ThreadSafeMutableRoaringBitmap queryableDocIds, Iterator<RecordInfo> recordInfoIterator) {
+    segment.enableUpsert(this, validDocIds, queryableDocIds);
```
[the remark this answers is not in the document]
Comment (tibrewalpratik17) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:169]: Good point!

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+    if (_deletedKeysTTL > 0) {
+      deletedKeysThreshold = largestSeenComparisonValue - _deletedKeysTTL;
+    } else {
+      deletedKeysThreshold = Double.MIN_VALUE;
+    }
+
+    _primaryKeyToRecordLocationMap.forEach((primaryKey, recordLocation) -> {
+      double comparisonValue = ((Number) recordLocation.getComparisonValue()).doubleValue();
+      if (_deletedKeysTTL > 0 && comparisonValue < deletedKeysThreshold
+          && recordLocation.getDistinctSegmentCount() <= 1) {
```
Comment (klsince) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:241]: perhaps add some comments on this check, or add comment on the disctinctSegmentCount field in  recordLocation that it's used here, as it's kinda critical for the correctness of this feature.

also looks like a lot of methods are same with ConcurrentMapPartitionUpsertMetadataManager. How about extracting a base class like BaseConcurrentMapPartitionUpsertMetadataManager, for both to reuse.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapTableUpsertMetadataManagerForConsistentDeletes.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.pinot.segment.spi.SegmentContext;
+
+
+/**
+ * Implementation of {@link TableUpsertMetadataManager} that is backed by a {@link ConcurrentHashMap} and ensures
+ * consistent deletions. This should be used when the table is configured with 'consistentDeletes' set to true.
+ */
+public class ConcurrentMapTableUpsertMetadataManagerForConsistentDeletes extends BaseTableUpsertMetadataManager {
```
Comment (klsince) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapTableUpsertMetadataManagerForConsistentDeletes.java:33]: Consider to extract a base class?

Comment (tibrewalpratik17) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapTableUpsertMetadataManagerForConsistentDeletes.java:33]: Other than `doUpdateRecord` and `getNumPrimaryKeys` none of the methods are same afaik.. since RecordLocation is a static class in each manager it makes it a bit unreadable when creating new RecordLocation in this class..

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java [hunk 10]
```diff
@@ -858,6 +858,38 @@ static void validateUpsertAndDedupConfig(TableConfig tableConfig, Schema schema)
       }
     }
 
+    // enableConsistentDeletes shouldn't exist with metadataTTL
+    if (upsertConfig != null && upsertConfig.isEnableConsistentDeletes()) {
```
Comment (klsince) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java:862]: why not combine all the checks inside one if-check? as the if (condition) is same

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/UpsertContext.java [hunk 11]
```diff
@@ -66,6 +69,8 @@ private UpsertContext(TableConfig tableConfig, Schema schema, List<String> prima
     _consistencyMode = consistencyMode;
     _upsertViewRefreshIntervalMs = upsertViewRefreshIntervalMs;
     _tableIndexDir = tableIndexDir;
+    _dropOutOfOrderRecord = dropOutOfOrderRecord;
```
Comment (klsince) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/UpsertContext.java:72]: is this change related to this PR or a quick fix?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/TableUpsertMetadataManagerFactory.java [hunk 12]
```diff
@@ -83,8 +82,14 @@ public static TableUpsertMetadataManager create(TableConfig tableConfig,
                 metadataManagerClass, tableNameWithType), e);
       }
     } else {
-      LOGGER.info("Creating ConcurrentMapTableUpsertMetadataManager for table: {}", tableNameWithType);
-      metadataManager = new ConcurrentMapTableUpsertMetadataManager();
+      if (upsertConfig.isEnableConsistentDeletes()) {
```
Comment (klsince) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/TableUpsertMetadataManagerFactory.java:85]: hmm.. if metadataManagerClass is set, the else-branch is skipped, so the check of the new feature flag is skipped too.

so should we check the feature flag check inside ConcurrentMapTableUpsertMetadataManager getOrCreatePartitionManager(), and based on the flag, we return either the original partition mgr, or the new partition mgr added in this PR.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+      _distinctSegmentCount = distinctSegmentCount;
+    }
+
+    public static int incrementSegmentCount(int count) {
+      return count + 1;
+    }
+
+    public static int decrementSegmentCount(int count) {
+      return count - 1;
+    }
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:373]: (minor) I don't think these 2 helper method is necessary. I found them less readable them simply `+1` or `-1`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+            return new RecordLocation(segment, newDocId, newComparisonValue, 1);
+          }
+        });
+
+    updatePrimaryKeyGauge();
+    return !isOutOfOrderRecord.get();
+  }
+
+  @Override
+  protected void removeSegment(IndexSegment segment, MutableRoaringBitmap validDocIds) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:323]: (minor) Do you need to override this? Making it no-op is confusing

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+    // Update metrics
+    long numPrimaryKeys = getNumPrimaryKeys();
+    updatePrimaryKeyGauge(numPrimaryKeys);
+    _logger.info("Finished removing segment: {} in {}ms, current primary key count: {}", segmentName,
+        System.currentTimeMillis() - startTimeMs, numPrimaryKeys);
+  }
+
+  @Override
+  protected void removeSegment(IndexSegment segment, Iterator<PrimaryKey> primaryKeyIterator) {
+    Set<Object> deDuplicatedPrimaryKeys = new HashSet<>();
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:208]: (minor)
```suggestion
    Set<Object> uniquePrimaryKeys = new HashSet<>();
```

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java [hunk 16]
```diff
@@ -0,0 +1,391 @@
...
+    while (primaryKeyIterator.hasNext()) {
+      PrimaryKey primaryKey = primaryKeyIterator.next();
+      _primaryKeyToRecordLocationMap.computeIfPresent(HashUtils.hashPrimaryKey(primaryKey, _hashFunction),
+          (pk, recordLocation) -> {
+            if (deDuplicatedPrimaryKeys.contains(pk)) {
+              return recordLocation;
+            }
+            deDuplicatedPrimaryKeys.add(pk);
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:216]: (minor)
```suggestion
            if (!uniquePrimaryKeys.add(pk)) {
              return recordLocation;
            }
```

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java (hunk 15 above)
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:208]: Please also add some comments explaining why we need to dedup it

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java [hunk 17]
```diff
@@ -0,0 +1,391 @@
...
+            if (recordLocation.getSegment() == segment) {
+              return null;
+            }
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:219]: (minor) Consider moving this check before unique key check to reduce key set size

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/TableUpsertMetadataManagerFactory.java (hunk 12 above)
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/TableUpsertMetadataManagerFactory.java:85]: Since the new `ConcurrentMapTableUpsertMetadataManagerForConsistentDeletes` is almost the same as `ConcurrentMapTableUpsertMetadataManager`, I guess we can move the logic into `ConcurrentMapTableUpsertMetadataManager.getOrCreatePartitionManager()` and simply change the value type of `_partitionMetadataManagerMap` to `BasePartitionUpsertMetadataManager`

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java [hunk 18]
```diff
@@ -79,6 +79,9 @@ public enum ConsistencyMode {
   @JsonPropertyDescription("Whether to preload segments for fast upsert metadata recovery")
   private boolean _enablePreload;
 
+  @JsonPropertyDescription("If we are using deletion + compaction we need to enable this for data consistency")
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:82]: This only associates with `_deletedKeysTTL`, so suggest moving it next to `_deletedKeysTTL` and make the name and description more specific. Currently people might misunderstand this to be associated with `_deleteRecordColumn`

Comment (tibrewalpratik17) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java]: Updated! Let me know if we can go for a better naming.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/TableUpsertMetadataManagerFactory.java (hunk 12 above)
Comment (tibrewalpratik17) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/TableUpsertMetadataManagerFactory.java]: Yes thanks Jackie for this suggestion! Looks neat now.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java (hunk 13 above)
Comment (tibrewalpratik17) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:371]: Hmm are you suggesting one helper method or no helper method at all and do +1, -1 inplace?

Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:371]: No helper method, simply `+1` and `-1` inplace

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java (hunk 18 above)
Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java]: Have you pushed the change?

Comment (tibrewalpratik17) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java]: Yes i renamed it to `_enableDeletedKeysCompactionConsistency`. It's updated in UpsertConfig class. Pushing a change to update in `UpsertContext` as well.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java (hunk 13 above)
Comment (tibrewalpratik17) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java:371]: can we do it in a follow-up? initially this will help me in debugging if we are missing out on any edge cases while doing addition / subtraction? 😅

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/TableUpsertMetadataManagerFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (klsince) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/TableUpsertMetadataManagerFactory.java:37]: looks like we don't need to change this class now that the flag is checked inside getOrCreatePartitionManager()?

**Review by klsince — APPROVED** [review]

**Comment by Jackie-Jiang** [comment]
Good job! Can you also help update the pinot doc to reflect this new feature?
